Diagnostics change for ambiguous overloads previously specifying 'this'

kathygray-pl opened the pull request:
Updates the message to use the type or ValueDecl instead of this on previous 'Found this candidate' messages Note: doees not yet change all test cases so more tests are failing

Improve Diagnostic message on overload ambiguitiy

Remove messages that say 'found this candidate' in favour of providing the type for the candidate to aid in selection when the candidates are presented in a dialogue window.

Fix more tests

<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

kathygray-pl committed:
Diagnostics change for ambiguous overloads previously specifying 'this'

Updates the message to use the type or ValueDecl instead of this on previous 'Found this candidate' messages
Note: doees not yet change all test cases so more tests are failing

Improve Diagnostic message on overload ambiguitiy

Remove messages that say 'found this candidate' in favour of providing the type for the candidate to aid in selection when the candidates are presented in a dialogue window.

Fix more tests

kathygray-pl committed:
Diagnostics change for ambiguous overloads previously specifying 'this'

Updates the message to use the type or ValueDecl instead of saying 'this' on 'Found this candidate' messages to provide more information where possible in selecting overloads

Per comment during draft iteration uses getInterfaceType and adds removeSelfParam to ValueDecl.t

Migrate to calling interfaceType and remove SelfParam

Update tests to reflect type information in ambiguity resolution

All but 5 tests now passing. Those 5 either do not refer to overloading errors or do pass even when modified to accept type specification

Move implementation to ValueDecl now that tests largely pass

kathygray-pl committed:
Correct missed with type phrase

kathygray-pl commented:
Although this isn't ready for review as there are more tests to bring into line (around 30), I'd love some early feedback on style and if this is the right track for my first change. Tagging those that look like the right owners based on code owners doc @hborla @slavapestov @hamishknight

hamishknight commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -4099,7 +4099,8 @@ bool SubscriptMisuseFailure::diagnoseAsError() {
   if (auto overload = getOverloadChoiceIfAvailable(getLocator())) {
-    emitDiagnosticAt(overload->choice.getDecl(), diag::found_candidate);
+    auto decl = overload->choice.getDecl();
+    emitDiagnosticAt(decl, diag::found_candidate_type, decl->getOverloadSignatureType());
```
You'll want to use `getInterfaceType` instead of `getOverloadSignatureType` (here and below), `getOverloadSignatureType` is what we use to determine if two decls overload each other, but may not actually have a type that reflects the decl, e.g all VarDecls have an overload type of `Void`.

I think we'll probably also want to drop the curried self from the type for type members, we probably ought to add a member to ValueDecl that returns this, the implementation would just be this: https://github.com/swiftlang/swift/blob/a5c6156525297716adefa58277d58d9ea8f6e63d/lib/Sema/AssociatedTypeInference.cpp#L2050-L2055

hamishknight reviewed (commented):
Nice! One comment about the type to print, but otherwise this LGTM

kathygray-pl committed:
Migrate to calling interfaceType and remove SelfParam

Migrate more tests

kathygray-pl committed:
Update tests to reflect type information in ambiguity resolution

All but 7 tests now passing. Those 7 tests either do not refer to these circumstances or do not seem to pass even when modified to accept type specifications.

asl commented:
autodiff test changes are obviously LGTM. Looks like a good addition!
